### ISSUE TITLE
Making network-based patient classifier scalable to larger datasets

### ISSUE THREAD

**Issue by shraddhapai** [nrnb/GoogleSummerOfCode#103]
### Background
**Why do people have different disease risk, different disease outcome, and difference responses to therapy? Can we use network-based algorithms and machine learning to identify features that distinguish different patient types?**

We have developed a supervised patient classifier that works by integrates heterogeneous clinical and cellular data types into a single model. For example, one could predict risk for lung cancer by integrating clinical, genetic, and transcriptomic data. The algorithm works by first converting each data type into a patient similarity network; in this network, nodes are patients and edges are weighted similarities between them. It then uses feature selection, a technique from the field of machine learning, to identify which inputs (networks) are predictive of a given patient class.

Our benchmarking results show that our method, **[netDx](http://netdx.org)**, is a top performer in predicting binary survival in four different cancers and provides insight into the biology of patient classes ([https://www.biorxiv.org/content/early/2017/11/10/084418]). However, we need netDx to scale to datasets with several thousand patients; datasets of this size are computationally intractable using our current implementation, because these generate a large memory footprint.

### Goal
Last year at GSoC, we implemented an alternative algorithm for network integration and feature selection (http://cb.csail.mit.edu/cb/mashup/). This implementation now needs to be tested and optimized against the original netDx implementation using benchmark datasets (examples include cancer survival from 5-6 datatypes; case/control prediction from genetic data). The goal is to optimize prediction accuracy, network weights for feature selection and test scalability. 

Our goal is to eventually replace the current feature selection algorithm underlying netDx with the implementation based on Mashup.

### Difficulty Level 3
This project requires a moderate understanding of matrix algebra (singular value decomposition). It requires the ability to write code to run new datasets and to troubleshoot errors by examining properties of the intermediate data (e.g. density plots, correlations, etc.,)

### Skills
* Strong programming skills in R, Python, Matlab and/or Julia (essential); the new algorithm is implemented in Julia so learning to work in this language is essential.
* Software development and testing expertise highly recommended
* Ability to generate plots and run simple descriptive and inferential statistics for troubleshooting is an asset (density plots, scatter plots, heatmaps, compute correlations). 

### Public Repository
The repository is not public as netDx is unpublished research. The repository will be made public upon publication of the methods paper.

### Potential Mentors
Gary Bader
Shraddha Pai
Shirley Hui

### Contact
[Shraddha Pai](mailto:shraddha.pai@utoronto.ca)

**Comment by bhaveshsunny** [comment]
Hi, I am bhavesh a GSOC'17 student of NRNB. I am interested in doing the project. I have worked on R, Python(keras,tensorflow,scikit-learn) and Matlab on a couple of analysis projects before. I am very interested in the project and would like your input on what you would expect in a potential student. Thanks.

**Comment by shraddhapai** [comment]
Hi Narra,
Thank you for getting in touch. I was traveling last week and have only
just been responding to emails.
It's great that you're interested in the project!

You will need to know basic machine learning concepts, be comfortable
working with Julia and R, and have good problem-solving and
software-testing skills. This project involves tuning an algorithm by
testing datasets, rather than software development of a user interface
(which is the upper layer). The portion in Julia runs a core process called
feature selection, and this "plugs into" the R layer which provides the
input/analyzes output/produces reports. You will be tweaking the Julia
layer and would be connecting the Julia to the R (although this is mostly
done), which is why you need to be somewhat familiar with both. We will be
testing different datasets on the algorithm and will focus heavily on
improving accuracy and scalability. The project will require simple
visualization and stats in R (boxplots, barplots with error bars, Venn
diagrams, t-tests, etc.,) on the results to convince ourselves that the
algorithm is working as intended, and we will develop a test suite to
examine results in various ways.

If you haven't done so already, read through the netDx manuscript (link in
GSoC project)  and see the work done on this project last year. The GSoC
'17 code repo is here, so you can see what the logic and Julia
implementation look like: https://github.com/memoiry/ModMashup.jl
The code currently runs and provides output but we need to optimize it to
match the accuracy of the current netDx core (which it is intended to
replace).

Does this help? Please don't hesitate to ask further questions, especially
where it pertains to understanding what the netDx algorithm does.
Also, is your plan to work remotely from India through the summer? You also
need to clarify what compute resources would be available to you. We may be
able to arrange something from our side but it is easier to have local
resources at your disposal; this is a compute-heavy project and you will
need machines with tens of Gigs of RAM.

Best,Shraddha
----
Shraddha Pai
Post-doctoral Fellow, http://baderlab.org
The Donnelly Center, University of Toronto

**Comment by sarthak-srivastava** [comment]
Hi, I am Sarthak and I am really interested in the project. I am comfortable in dealing with machine learning problems in python,R (standard libraries),MATLAB and have done  projects on same. I have started to familiarize myself with Julia. Cloud computing could be a potential solution to the hardware resources problem.

**Comment by bhaveshsunny** [comment]
@shraddhapai  I have read the algorithm and I have a couple of things I would like to discuss with you. I have also tried to get the Mashup workflow running on my pc. I have contacted you by email but I haven't received a reply.

**Comment by shraddhapai** [comment]
Hi @bhaveshsunny, I've responded to your previous email in the forum above so that the response would be useful to other candidates as well - have you had a chance to go through that? I have received your email from yesterday as well and we can continue specific correspondence that way.
Regards,Shraddha

**Comment by shraddhapai** [comment]
Hi @sarthak-srivastava, great that you are interested in the project. Let me know if you have any questions as you decide whether to pursue this project. We are based in Toronto; will you be working remotely? Best,Shraddha

**Comment by utkarshshukla2912** [comment]
Hi, I am Utkarsh and I am really interested in the project. I have a considerable experience with different machine learning as well as deep learning algorithms which comes from my previous research and projects in the field. My work has focused on BCI (BRAIN COMPUTER INTERFACE) and EEG data. I have been affiliated with medical data for a while and hope to contribute to the project.
Thanks

**Comment by shraddhapai** [comment]
Hi,
I've received questions about what we expect in the GSoC proposal for the year.
If you haven't already done so, please read my comments above in this thread. 

We already have an implementation of Mashup in Julia (GSoC 2017). We also already have a branch of the netDx R code that uses this algorithm for feature selection. The goals for this summer will be to get this new implementation to the point where it becomes the standard version we use. 

Specific goals are:
1) to measure the accuracy of this version (mashup-netDx) by comparing the feature scores and classification accuracy to the current version of netDx. We will run the new classifier on several example datasets that have been used for the current netDx prediction, including case studies in breast cancer, asthma and schizophrenia. We will want to tweak the current implementation till we achieve the desired functionality. 

2) to compare the scalability of mashup-netDx to the current version. Scalability to EHR-scale genomic data is the main reason we are hoping to switch to mashup-netDx. We will build predictors that integrate clinical and genetic SNP data, with examples including rheumatoid arthritis, Alzheimer's disease, and human cognition. 

--- 
* Your application needs to demonstrate that you have a general understanding of netDx works, and a general understanding of how the Mashup algorithm implemented last summer works. A good way to do this is to provide a brief summary of project background at the top of your application.
* Tell us what experience you have that is relevant and where you would draw on it in the current project. Be specific.
    Generic: "machine learning" or "Python/R" 
    Specific: Discussing a classifier you worked on, genomic data analysis you did, testing you did, and how that would flow into this project. 
* We are looking for experience in different areas including: 
   - data analysis and machine learning; 
   - developing software for analysis; 
   - ability to test correctness of software; 
   - clear communication skills (documentation is key!); 
   - ability to work well with moderate-to-minimal supervision but ask for guidance when necessary. 
   - Being self-driven is an important asset. So is being able to work with critical feedback.
Your application should demonstrate that you have these skills. Again, specific examples.
* Provide a rough timeline of how you would achieve the goals above. It doesn't have to be detailed. We are looking to see a candidate's ability to plan out a project. The actual timelines will change in the course of the project.

Please ask in this forum if you have more questions. Do not e-mail me as others won't learn from the exchange. 

Best,Shraddha

**Comment by bhaveshsunny** [comment]
@shraddhapai  Thanks for the detailed response.

**Comment by narendrameena** [comment]
@shraddhapai thanks for the detailed info that clarifies a lot.

**Comment by bhaveshsunny** [comment]
@shraddhapai  Could you  add the dataset here? That would help us a lot! thanks again.

**Comment by shraddhapai** [comment]
Hello GSoC applicants,
Here is a dataset you may want to try to use modMashup with for classification. 
The task is binary classification of a breast tumour as resulting in good or poor survival. The data provided includes gene expression and a few clinical parameters. netDx groups genes by their pathways. 

This Rdata contains all the information you need: 
https://drive.google.com/open?id=19c3VDKH1uvZf7PDzQbPYv84IflwQymiN

## Load data file in R
```
$ R
> load("BRCA_demo.Rdata")
```

## Phenotype
This is the sample metadata table. Patient identifiers are in the "ID" column and the label to classify is in the "STATUS" column.
```
> head(pheno)
            ID age grade     stage STATUS_INT     STATUS
1 TCGA-AK-3428  62    G2 Stage III          1 SURVIVEYES
2 TCGA-AK-3434  72    G2   Stage I          1 SURVIVEYES
3 TCGA-B0-4688  46    G4  Stage IV          0  SURVIVENO
4 TCGA-B0-4690  65    G3  Stage IV          0  SURVIVENO
5 TCGA-B0-4691  55    G3  Stage IV          0  SURVIVENO
6 TCGA-B0-4693  72    G4 Stage III          0  SURVIVENO
```

## Gene expression data
Standard tabular format with genes in rows and patients in columns.
```
> head(dats$rna[,1:5])
        TCGA-AK-3428 TCGA-AK-3434 TCGA-B0-4688 TCGA-B0-4690 TCGA-B0-4691
ZNF121       72.1289      47.3807      64.2683      81.1642         22.5
OR2J3         0.0000       0.0000       0.0000       0.0000          0.5
HMOX1     10726.8908    3753.7897     675.0007    6300.2797       7766.0
SYT4          0.0000       0.0000       0.0000       0.0000          2.5
GPR137C      39.9160      38.3558     409.1136      30.5297         17.5
AKAP12     5680.6723    2663.4704    1041.5136    3684.7812       3173.5
```

## Grouping information
We would create one input patient similarity network per group (pathway for genes ; one net for all clinical data).
```
> summary(groupList)
         Length Class  Mode
rna      1801   -none- list
clinical    3   -none- list
> head(groupList$rna)
$GUANOSINE_NUCLEOTIDES__I_DE_NOVO__I__BIOSYNTHESIS
 [1] "NME7"   "NME6"   "RRM2B"  "GMPS"   "NME2"   "NME3"   "NME4"   "NME5"
 [9] "RRM2"   "NME1"   "GUK1"   "RRM1"   "IMPDH2" "IMPDH1"

$RETINOL_BIOSYNTHESIS
 [1] "RDH10" "DHRS4" "LRAT"  "LIPC"  "CES5A" "DHRS9" "RDH11" "DHRS3" "CES1"
[10] "RBP1"  "CES4A" "RBP2"  "PNLIP" "RBP5"  "RBP4"  "CES2"

$`MUCIN_CORE_1_AND_CORE_2__I_O__I_-GLYCOSYLATION`
 [1] "GALNT1"  "GCNT4"   "GALNT7"  "GCNT3"   "GCNT7"   "GALNT6"  "GALNT4"
 [8] "GALNT5"  "ST3GAL2" "ST3GAL1" "ST3GAL4" "GALNT10" "GALNT15" "GALNTL6"
[15] "B3GNT3"  "GALNT16" "GALNT18" "GALNT11" "GALNT12" "GCNT1"   "C1GALT1"
[22] "GALNT13" "GALNT14" "WBSCR17" "GALNT8"  "GALNT9"  "GALNT2"  "GALNT3"

$`SUPERPATHWAY_OF_D-_I_MYO__I_-INOSITOL__1,4,5_-TRISPHOSPHATE_METABOLISM`
 [1] "MINPP1" "INPP5A" "INPP5B" "INPP5K" "INPP5J" "ITPKA"  "PTEN"   "INPP5D"
 [9] "IPMK"   "SYNJ2"  "INPP5F" "IMPA1"  "INPP1"  "INPPL1" "IMPA2"  "IMPAD1"
[17] "ITPKC"  "OCRL"   "SYNJ1"  "ITPKB"

$`D-_I_MYO__I_-INOSITOL__1,4,5_-TRISPHOSPHATE_DEGRADATION`
 [1] "INPP5A" "INPP5B" "INPP5K" "INPP5J" "SYNJ2"  "INPP5F" "IMPA1"  "INPP1"
 [9] "INPPL1" "IMPA2"  "IMPAD1" "OCRL"   "SYNJ1"

$MRNA_CAPPING
 [1] "CCNH"   "RNGTT"  "MNAT1"  "GTF2F1" "GTF2F2" "SUPT5H" "ERCC2"  "ERCC3"
 [9] "POLR2A" "POLR2B" "POLR2C" "POLR2D" "POLR2I" "POLR2J" "POLR2K" "POLR2L"
[17] "POLR2E" "POLR2G" "POLR2F" "NCBP2"  "POLR2H" "RNMT"   "NCBP1"  "GTF2H2"
[25] "GTF2H1" "CDK7"   "GTF2H3" "GTF2H5" "GTF2H4"
```

The goal would be to run modMashup and get weights for each network, with greater weights indicating higher predictive value.

Shraddha
